pkg/frame/native.go

```diff
@@ -1,62 +1,195 @@
 package frame
 
 import (
+	"errors"
+	"fmt"
 	"image"
 	"image/color"
+
+	"golang.org/x/exp/constraints"
 )
 
+var ErrUnsupportedSamplesPerPixel = errors.New("unsupported samples per pixel")
+
+// INativeFrame is an interface representation of NativeFrame[I]'s capabilities,
+// and offers a way to use a NativeFrame _without_ requiring propogation of
+// type parameters. This allows for some more ergonomic signatures, though
+// NativeFrame[I] can be used directly as well for those who prefer it.
+type INativeFrame interface {
+	// Rows returns the number of rows in this frame (which is the max y
+	// dimension).
+	Rows() int
+	// Cols returns the number of columns in this frame (which is the max x
+	// dimension).
+	Cols() int
+	// SamplesPerPixel returns the number of samples per pixel in this frame.
+	SamplesPerPixel() int
+	// BitsPerSample returns the bits per sample in this frame.
+	BitsPerSample() int
+	// GetPixel returns the samples (as a slice) for the pixel at (x, y).
+	// The coordinate system of the image starts with (0, 0) in the upper left
+	// corner of the image, with X increasing to the right, and Y increasing
+	// down:
+	//
+	//  0 -------▶ X
+	//  |
+	//  |
+	//  ▼
+	//  Y
+	GetPixel(x, y int) ([]int, error)
+	// RawDataSlice will return the underlying data slice, which will be of type
+	// []I. Based on BitsPerSample, you can anticipate what type of slice you'll
+	// get, and type assert as needed:
+	//  BitsPerSample    Slice
+	//  8                []uint8
+	//  16               []uint16
+	//  32               []uint32
+	RawDataSlice() any
+	// Equals returns true if this INativeFrame exactly equals the provided
+	// INativeFrame. This checks every pixel value, so may be expensive.
+	// In the future we may compute a one time hash during construction to make
+	// this less expensive in the future if called multiple time.
+	Equals(frame INativeFrame) bool
+	CommonFrame
+}
+
 // NativeFrame represents a native image frame
-type NativeFrame struct {
-	// Data is a slice of pixels, where each pixel can have multiple values
-	Data          [][]int
-	Rows          int
-	Cols          int
-	BitsPerSample int
+type NativeFrame[I constraints.Integer] struct {
+	// RawData is a slice of pixel values. For each pixel, each sample for the
+	// pixel is unrolled per pixel. For example, consider 2 pixels that have 3
+	// samples per Pixel: [[1,2,3], [4,5,6]]. This would be unrolled like
+	// [1,2,3,4,5,6]. The pixels themselves are arranged in row order, so the
+	// first row of pixels would be unrolled in order, followed by the next row,
+	// and so on in this flattened array.
+	// A flattened slice is used instead of a nested 2D slice because there is
+	// significant overhead to creating nested slices in Go discussed here:
+	// https://github.com/suyashkumar/dicom/issues/161#issuecomment-2143627792.
+	RawData                 []I
```

Comment:
Assuming one day, go have native support for JPEG2000, how you will adapt it into this project.

How we can allow users to decode DICOM, alter the image, and then write back to another DICOM files.

For the sustainable development of this library, do you think the data structure + interface are roburst to changes in the future.

-----
Assuming there is a developer who want to write a plugin that decodes JPEG2000. How can we make our API open to that need.

Reply:
> Assuming one day, go have native support for JPEG2000, how you will adapt it into this project.

This is the NativeFrame construct, so this is only for representing Native PixelData that is encoded in binary format as various sizes of integers. And specifically, this is the container we parse that data into from the dicom. Encapsulated images encoded in JPEG2000 or other flavors of JPEG will not be represented directly in this format.

This is worth thinking about some more, but in the current API, any decoding (including JPEG/JPEG2000) of raw data would occur in [EncapsulatedFrame.GetImage()](https://github.com/suyashkumar/dicom/blob/cb3864efadad3b9fd1dc331f5763bb960bc79bad/pkg/frame/encapsulated.go#L35) which is the abstraction layer function for returning a standard Golang `image.Image`. However, image.Image may itself be an imperfect representation as it may be lossy in terms of precision for some kinds of image data that dicoms can represent to higher precision. So this is something we should dig into a bit deeper imo. 

> Assuming there is a developer who want to write a plugin that decodes JPEG2000. How can we make our API open to that need.

Very good question, as I know you're thinking a bit about this. I'm very open to thoughts and suggestions btw. The first thing we need to decide: is `image.Image` a suitable decoding target? If yes, then the answer would be that a contribution should be made to add any peeking + decoding logic to [EncapsulatedFrame.GetImage()](https://github.com/suyashkumar/dicom/blob/cb3864efadad3b9fd1dc331f5763bb960bc79bad/pkg/frame/encapsulated.go#L35). If we want to allow users to register their own decoders, we can call the appropriate decoding logic here in this function based on global registrations seen so far. 

One idea that might be worth thinking about--should we have a better intermediate decoding target other than image.Image? e.g. some structure holding `[]T` along with some rows, cols, etc (something similar to NativeFrame)? 

> For the sustainable development of this library, do you think the data structure + interface are roburst to changes in the future.

I think this is a reasonable way to represent Native PixelData internally, mostly because this is how Native PixelData is modeled in a DICOM, and now we've correctly allowed for generic types to be used for storage (though we can expand the generic envelope here some more for signed ints, float PixelData, etc). Most folks will not need to interact with this directly though, and will interact via the NativeFrame's helper methods, where we can add additional layers of abstraction and ergonomics as needed. What do you think? Very open to changes and feedback, and don't think this is a final version or anything :). 

> How we can allow users to decode DICOM, alter the image, and then write back to another DICOM files.

This workflow needs improvement for sure. This is possible to do today very manually, but I definitely think we should allow for helper methods/packages/examples as makes sense to help with key parts of this process. Right now, you can call `GetImage()` on any `frame.CommonFrame` and get an `image.Image`. You can then manipulate that image.Image however you see fit, then encode it (manually, using system libraries) to binary bytes and drop it back into an EncapsulatedFrame, then write it out later. Some pieces here are missing, including setting the transfer syntax for the updated image if it's done in a different way. 

Some thoughts on TODOs:
* Determine if we want image.Image to be the Golang abstraction for working with images
* Consider storing the transfer syntax uid on frames internally, and consult that later when writing the data back out
* Write helper function to: `ToFrame(img image.Image, outputTransferSyntaxUID string, frameMetadata ...) frame.Frame` (the inverse of GetImage). Users can call something like this to create a new frame that can be added to a PixelDataInfo to be written out in a Dataset. Maybe we can have a common frameMetadata structure shared across frames to make life easy. Users would still need to take some care to edit other associated tags if needed for their use case (e.g. windowWidth, windowlevel, etc). 

Anyway sorry for the long comment! Just some off the cuff thoughts.